In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [3]:
df = pd.read_csv('/Users/namhunkim/Library/CloudStorage/OneDrive-연세대학교(YonseiUniversity)/DSL/EDA/data/youtube_0717.csv', lineterminator='\n')
df.head(2)

video_id               title          publishedAt  \
0  uq5LClQN3cE         안녕하세요 보겸입니다  2020-08-09 09:32:48   
1  I-ZbZCHsHD0  부락토스의 계획 [총몇명 프리퀄]  2020-08-12 09:00:08   

                  channelId channelTitle        trending_date  \
0  UCu9BCtGIEr73LXZsKmoujKw        보겸 BK  2020-08-12 00:00:00   
1  UCRuSxVu4iqTK5kCh90ntAgA          총몇명  2020-08-12 00:00:00   

                                  tags  view_count  likes  dislikes  \
0                            보겸|bokyem     5947503  53326    105756   
1  총몇명|재밌는 만화|부락토스|루시퍼|총몇명 프리퀄|총몇명 스토리      963384  28244       494   

   comment_count  comments_disabled  ratings_disabled  \
0         139946              False             False   
1           3339              False             False   

                                         description     category_name  \
0                                                NaN     Entertainment   
1  오늘도 정말 감사드립니다!!총몇명 스튜디오 - 총몇명, 십제곱, 5G민, MOVE혁...  Film & Animation   

   tag_num  
0        2  
1        6

In [4]:
df.columns

Index(['video_id', 'title', 'publishedAt', 'channelId', 'channelTitle',
       'trending_date', 'tags', 'view_count', 'likes', 'dislikes',
       'comment_count', 'comments_disabled', 'ratings_disabled', 'description',
       'category_name', 'tag_num'],
      dtype='object')

# 기존 데이터 분석

## 1. 제목길이, 채널길이, description길이, tag길이

In [5]:
x = df.iloc[12,:]
x

video_id                                                   rRaBKB9gDSQ
title                                      뒷광고 논란에 대한 해명 및 전할 말씀이 있습니다
publishedAt                                        2020-08-09 16:12:59
channelId                                     UCBIoXzDldCnpbM_7uyG0_Tg
channelTitle                                                HONG SOUND
trending_date                                      2020-08-12 00:00:00
tags                 ASMR|홍사운드|eatingsound|eating|리얼사운드|이팅사운드|먹는소리|...
view_count                                                     2519737
likes                                                            71721
dislikes                                                          2863
comment_count                                                    16283
comments_disabled                                                False
ratings_disabled                                                 False
description          인스타그램(instagram) : https://www.instagram.com/h...
catego

In [6]:
print('title :', x.title)
print('channelTitle :', x.channelTitle)
print('description :', x.description)
print('tags :', x.tags)

title : 뒷광고 논란에 대한 해명 및 전할 말씀이 있습니다
channelTitle : HONG SOUND
description : 인스타그램(instagram) : https://www.instagram.com/hong_sound
tags : ASMR|홍사운드|eatingsound|eating|리얼사운드|이팅사운드|먹는소리|리얼먹방|social mukbang|mukbang|eatingshow|먹방|yt:cc=on


In [6]:
# all characters should be transformed into number

print('title :', len(x.title))
print('channelTitle :', len(x.channelTitle))
print('description :', len(x.description))
print('tags :', x.tag_num)

title : 27
channelTitle : 10
description : 55
tags : 13


### 1-(1). 상관계수

In [7]:
# 1. Using 'df.corr()'

df_len = df.loc[:,['title','channelTitle','description','tag_num']]
df_len_converted = df_len

df_len_converted['title'] = df_len['title'].apply(lambda x : len(x))
df_len_converted['channelTitle'] = df_len['channelTitle'].apply(lambda x : len(x))
# description의 경우에는 NaN값이 존재한다.
df_len_converted['description'] = df_len['description'].apply(lambda x : len(x) if x==x else 0)

print(df_len_converted.head(5))
print('-'*50)
print(df_len_converted.corr())

   title  channelTitle  description  tag_num
0     11             5            0        2
1     18             3          496        6
2     15            11            0       27
3     20            13           35       12
4     48            17          524       13
--------------------------------------------------
                 title  channelTitle  description   tag_num
title         1.000000      0.210965     0.124431  0.259984
channelTitle  0.210965      1.000000     0.123917  0.112540
description   0.124431      0.123917     1.000000  0.066646
tag_num       0.259984      0.112540     0.066646  1.000000


In [8]:
# 2. Using s1.corr(s2)

column = df_len_converted.columns
x = df_len_converted

for i in range(len(column)):
    for j in range(len(column)):
        if i < j:
            print('{0} & {1} : {2}'.format(column[i],column[j],
                                           np.round((x[column[i]]).corr(x[column[j]]), 2)))

title & channelTitle : 0.21
title & description : 0.12
title & tag_num : 0.26
channelTitle & description : 0.12
channelTitle & tag_num : 0.11
description & tag_num : 0.07


In [9]:
# 3. Using np.corrcoef(ndarray1, ndarray2)

for i in range(len(column)):
    for j in range(len(column)):
        if i < j:
            print('{0} & {1} : {2}'.format(column[i],column[j],
                                           np.round(np.corrcoef((x[column[i]]).values, (x[column[j]]).values)[0,1], 2)))

title & channelTitle : 0.21
title & description : 0.12
title & tag_num : 0.26
channelTitle & description : 0.12
channelTitle & tag_num : 0.11
description & tag_num : 0.07


In [10]:
# Conclusion : 가장 높은 상관계수값이 0.26으로 길이간의 상관성은 없다는 결론에 힘이 실린다.
# Observation : description의 경우에는 0(NaN)이 다수 포함되기에 descrition과의 상관계수들은 모두 낮게 나온다.
# 이는 tag_num도 마찬가지, 반면에 제목과 채널이름은 0(NaN)이 될 수 없음으로 상대적으로 높게 나온다.

for i in column:
    print(i, ':', x[i][x[i]==0].count()) # same as len(np.where(x.description==0)[0])

title : 0
channelTitle : 0
description : 3543
tag_num : 18576


In [11]:
# 1. drop를 통해 제거 <- 시간이 꽤 걸린다. 왜 이렇게 오래 걸리는지는 모르겠다.

x1 = df_len_converted

for i in x1.index:
    if (x1.loc[i,:]==0).any(): # 0이 하나라도 있는 행에 대하여 제거
        x1 = x1.drop(i)
x1.head(5)

KeyboardInterrupt: 

In [12]:
# 2. concat을 통해 합치기 <- 마찬가지로 좀 오래 걸린다.

x = df_len_converted
x2 = pd.DataFrame(np.zeros(4).reshape(1,4), columns=x.columns) # 붙일 장소

for i in x.index:
    if (x.loc[i,:]!=0).all(): # 0이 하나도 없는 행에 대하여 추가
        x_temp = pd.DataFrame(x.loc[i,:].values.reshape(1,4), index=[i], columns=x.loc[1,:].index)
        x2 = pd.concat([x2, x_temp], axis=0)

x2 = x2.drop(0) # 이전 0번째 삭제        
x2.head(5)

title  channelTitle  description  tag_num
1   18.0           3.0        496.0      6.0
3   20.0          13.0         35.0     12.0
4   48.0          17.0        524.0     13.0
5   25.0          17.0        301.0     47.0
8   29.0          11.0        810.0     67.0

In [226]:
# 0이 존재하는 행은 모두 삭제되었다.

for i in column:
    print(i, ':', x1[i][x1[i]==0].count()) # same as len(np.where(x.description==0)[0])
print('-'*50, '\n')

# 두 방식 모두 동일한 값을 도출한다.

print((x1 == x2).all(), '\n')
print(x1.shape)
print(x2.shape)

title : 0
channelTitle : 0
description : 0
tag_num : 0
-------------------------------------------------- 

title           True
channelTitle    True
description     True
tag_num         True
dtype: bool 

(117813, 4)
(117813, 4)


In [13]:
x2.head(5)

title  channelTitle  description  tag_num
1   18.0           3.0        496.0      6.0
3   20.0          13.0         35.0     12.0
4   48.0          17.0        524.0     13.0
5   25.0          17.0        301.0     47.0
8   29.0          11.0        810.0     67.0

In [14]:
# 따라서 0(NaN)을 제거하고 상관관계를 다시 한 번 구해보자.

column = x2.columns

print('[new]\n')
for i in range(len(column)):
    for j in range(len(column)):
        if i < j:
            print('{0} & {1} : {2}'.format(column[i],column[j],
                                           np.round((x2[column[i]]).corr(x2[column[j]]), 2)))
            
print('-'*50, '\n')
print('[old]\n')

column = df_len_converted.columns
x = df_len_converted

for i in range(len(column)):
    for j in range(len(column)):
        if i < j:
            print('{0} & {1} : {2}'.format(column[i],column[j],
                                           np.round((x[column[i]]).corr(x[column[j]]), 2)))

[new]

title & channelTitle : 0.2
title & description : 0.09
title & tag_num : 0.23
channelTitle & description : 0.12
channelTitle & tag_num : 0.09
description & tag_num : 0.03
-------------------------------------------------- 

[old]

title & channelTitle : 0.21
title & description : 0.12
title & tag_num : 0.26
channelTitle & description : 0.12
channelTitle & tag_num : 0.11
description & tag_num : 0.07


### 결론
*  오히려 0값을 제거하니 상관관계가 더 떨어졌다.
* 따라서 제목과 채널이름 그리고 태그와 묘사들의 길이 간에는 상관관계가 없다고 판단되었다.
* 그나마 높은 항목은 (제목,테그)=0.26, (제목,채널이름)=0.21 이다.

### 1-(2). 인기동영상을 지속적으로 배출하는 그룹(group1)과 그렇지 않은 그룹(group2)간에 좋아요나 싫어요 등에 유의미한 차이가 있는가? (조회수에 대비하여)

In [15]:
df.head(1)

video_id        title          publishedAt                 channelId  \
0  uq5LClQN3cE  안녕하세요 보겸입니다  2020-08-09 09:32:48  UCu9BCtGIEr73LXZsKmoujKw   

  channelTitle        trending_date       tags  view_count  likes  dislikes  \
0        보겸 BK  2020-08-12 00:00:00  보겸|bokyem     5947503  53326    105756   

   comment_count  comments_disabled  ratings_disabled description  \
0         139946              False             False         NaN   

   category_name  tag_num  
0  Entertainment        2

#### 채널이름과 아이디는 서로 일대일 대응인가?

In [16]:
# 채널이름과 아이디는 서로 일대일 대응인가?

print('# of unique channelId :', len(df.channelId.unique())) # == len(np.unique(df.channelId))
print('# of unique channelTitle :', len(df.channelTitle.unique()))

# 일대일 대응이 아니다. 채널 제목의 수가 200개 정도 더 많다.
# 아마 채널이름을 변경하면 채널아이디는 고정이 되는 듯 하다.
# 즉, 채널이름이 여러개인 채널아이디가 존재할 것
# 반대로 채널아이디가 여러개인 채널이름은 없을 것이다.

# of unique channelId : 2868
# of unique channelTitle : 3038


In [8]:
x = df.loc[:,['channelId','channelTitle']]

print('채널아이디가 동일하나 채널명이 다른 것들의 갯수 : ', (x.groupby('channelId').nunique() > 1).sum().values[0])
print('채널명이 같으나 채널아이디가 다른 것들의 갯수 : ', (x.groupby('channelTitle').nunique() > 1).sum().values[0])

채널아이디가 동일하나 채널명이 다른 것들의 갯수 :  172
채널명이 같으나 채널아이디가 다른 것들의 갯수 :  5


In [9]:
# 채널명이 같으나 채널아이디가 다른 것들을 찾아보자.

print('index :', np.where(x.groupby('channelTitle').nunique()>1)[0], '\n')
print(x.groupby('channelTitle').nunique().iloc[np.where(x.groupby('channelTitle').nunique()>1)[0],:])

index : [ 99 435 586 789 882] 

                         channelId
channelTitle                      
BE'O - Topic                     2
KONNECT Entertainment            2
NCT U - Topic                    2
THE BOYZ                         2
Various Artists - Topic         27


In [10]:
# 왜 중복이 되는지는 모르겠으나 확실히 다른 아이디를 가졌다.
# 계정을 동일한 채널명으로 또 팠을수도 있다.

df[df.channelTitle=='BE\'O - Topic'].channelId.unique()

array(['UC1B59rS0Gf9loL9ggfMSpfA', 'UCZLIclg1_Jgrrh8lBgP2mXQ'],
      dtype=object)

In [11]:
# 채널아이디가 같으나 채널명이 다른 것들을 찾아보자.

print('index :', np.where(x.groupby('channelId').nunique()>1)[0], '\n')
print(x.groupby('channelId').nunique().iloc[np.where(x.groupby('channelId').nunique()>1)[0],:])

index : [  18   20   34   57   63   75   79  155  161  200  203  239  244  256
  270  274  282  300  305  311  312  359  416  420  422  429  435  473
  501  508  521  557  562  575  577  597  598  607  627  669  680  696
  716  726  741  761  790  792  835  836  843  848  849  874  877  900
  922  939  980  997 1031 1059 1072 1098 1111 1161 1201 1203 1210 1238
 1244 1260 1267 1272 1286 1338 1351 1371 1382 1399 1421 1453 1461 1474
 1508 1567 1591 1609 1612 1642 1648 1669 1712 1724 1765 1802 1812 1864
 1875 1890 1905 1907 1915 1931 1973 1994 2015 2029 2035 2039 2046 2057
 2059 2089 2098 2102 2109 2142 2160 2161 2171 2172 2182 2188 2189 2216
 2218 2237 2246 2267 2272 2285 2302 2309 2318 2325 2326 2336 2337 2349
 2381 2386 2412 2436 2449 2470 2481 2518 2554 2560 2569 2575 2587 2601
 2613 2618 2632 2645 2650 2655 2705 2749 2757 2766 2782 2803 2827 2830
 2832 2833 2848 2853] 

                          channelTitle
channelId                             
UC-IBt8pM8hWx8wiwjcDLdIQ             2

In [12]:
y = x.groupby('channelId').nunique().iloc[np.where(x.groupby('channelId').nunique()>1)[0],:].channelTitle

print('채널아이디별 중복되는 채널이름의 갯수 :', y.unique())

채널아이디별 중복되는 채널이름의 갯수 : [2 3 4]


In [13]:
y = x.groupby('channelId').nunique().iloc[np.where(x.groupby('channelId').nunique()>1)[0],:]['channelTitle']
y.value_counts()

2    147
3     22
4      3
Name: channelTitle, dtype: int64

In [14]:
len(list(y.index))

172

#### 결론
* 채널명보다 채널아이디의 갯수가 더 적다.
* 같은 채널명을 가지지만 채널아이디가 다른 경우는 거의 없다.(5개)
* 반대로 같은 채널아이디에 다른 채널명인 경우는 172개이지만 전체갯수에 비해서는 적다.
* 크게 중요하지는 않지만 그래도
* 채널명이 아닌 채널아이디를 기준으로 그룹을 나누어보자.
* group1 = 인기동영상이 1개인 채널 ()
* group2 = 인기동영상이 2개이상인 채널 (147+22+3=172)

In [15]:
df.columns

Index(['video_id', 'title', 'publishedAt', 'channelId', 'channelTitle',
       'trending_date', 'tags', 'view_count', 'likes', 'dislikes',
       'comment_count', 'comments_disabled', 'ratings_disabled', 'description',
       'category_name', 'tag_num'],
      dtype='object')

In [16]:
x = df.loc[:,['video_id','channelId']]
x.head(1)

video_id                 channelId
0  uq5LClQN3cE  UCu9BCtGIEr73LXZsKmoujKw

In [17]:
# group2 (채널아이디 하나에 2개이상의 비디오아이디가 있는 그룹)

y = x.groupby('channelId').nunique().iloc[np.where(x.groupby('channelId').nunique()>1)[0],:]
idx = y.index
idx # 인덱스 추출

Index(['UC-0229MiVYnQoUxdLJnFl_g', 'UC-0C8yVGJy-cS4FGlYKelWw',
       'UC-0tICWyA0_AJAVMBHL8GdQ', 'UC-5acDjU6XFWu-__G17V_YQ',
       'UC-603KNB_fvzHP9MpBqSKWg', 'UC-60AlFBYJEoItr-u-tQXbw',
       'UC-7ZO1RZQq4O_xq9Zu8CyOg', 'UC-9RQCJ0gS_Kgt7J0SVrE8w',
       'UC-BqPABOl7c7rR5EoJJZ0UA', 'UC-Bsa2ivAGWq7bsSPrPGFVA',
       ...
       'UCzgxx_DM2Dcb9Y1spb9mUJA', 'UCzh4yY8rl38knH33XpNqXbQ',
       'UCzjDgV-G8_J-iRbkYktqyGQ', 'UCzjPaOBCe0K2ma8XJzsNonA',
       'UCzjiRWjfbtHwdpSLTs0LzUw', 'UCznImSIaxZR7fdLCICLdgaQ',
       'UCzt24ffdUiJvJAm5HY084aw', 'UCzteSXznVjwESizsCcSKSWw',
       'UCzw-C7fNfs018R1FzIKnlaA', 'UCzz58-H0wzmGoFi3MO_9ePQ'],
      dtype='object', name='channelId', length=1805)

In [18]:
# 프로그램이 완료될 때까지 시간이 꽤 소요된다.

n = len(df.columns) # 16
group1 = pd.DataFrame(np.zeros(n).reshape(1,n), columns=df.columns) # 붙일 장소
group2 = pd.DataFrame(np.zeros(n).reshape(1,n), columns=df.columns)

for i in df.index:
    if (df.loc[i,'channelId'] in idx): # 2개이상의 인기동영상을 가진 채널이면
        x_temp1 = pd.DataFrame(df.loc[i,:].values.reshape(1,n), index=[i], columns=df.loc[1,:].index)
        group2 = pd.concat([group2, x_temp1], axis=0)
    else: # 1개의 인기동영상만을 가진 채널이면
        x_temp2 = pd.DataFrame(df.loc[i,:].values.reshape(1,n), index=[i], columns=df.loc[1,:].index)
        group1 = pd.concat([group1, x_temp2], axis=0)

group1 = group1.drop(0) # 이전 0번째 삭제
group2 = group2.drop(0)

In [19]:
group1.head(1)

video_id                  title          publishedAt  \
24  oQHdEeUcfO8  룬테라 이야기: 타곤 | '도약의 길'  2020-08-11 15:00:13   

                   channelId channelTitle        trending_date  \
24  UCLuCG6sPAQX_bhTt1ztN1FQ   레전드 오브 룬테라  2020-08-12 00:00:00   

                                                 tags view_count likes  \
24  Riotgames|라이엇 게임즈|Riot|라이엇|리그오브레전드|리그 오브 레전드|롤...     158938  1077   

   dislikes comment_count comments_disabled ratings_disabled  \
24       20           554             False            False   

                                          description category_name tag_num  
24  타곤의 루나리와 솔라리에 합류하시겠어요? 지금 바로 https://playrunet...        Gaming      18

In [20]:
group2.head(1)

video_id               title          publishedAt  \
1  I-ZbZCHsHD0  부락토스의 계획 [총몇명 프리퀄]  2020-08-12 09:00:08   

                  channelId channelTitle        trending_date  \
1  UCRuSxVu4iqTK5kCh90ntAgA          총몇명  2020-08-12 00:00:00   

                                  tags view_count  likes dislikes  \
1  총몇명|재밌는 만화|부락토스|루시퍼|총몇명 프리퀄|총몇명 스토리     963384  28244      494   

  comment_count comments_disabled ratings_disabled  \
1          3339             False            False   

                                         description     category_name tag_num  
1  오늘도 정말 감사드립니다!!총몇명 스튜디오 - 총몇명, 십제곱, 5G민, MOVE혁...  Film & Animation       6

In [25]:
# 조회수 평균비교

print('view count Comparision\n')
print('group1 : {0} vs group2 {1}'.format(np.round(group1.view_count.mean(),0),
                                          np.round(group2.view_count.mean(),0)))

view count Comparision

group1 : 1318959.0 vs group2 1637863.0


In [29]:
# 좋아요수,싫어요수 평균비교

print('likes Comparision\n')
print('group1 : {0} vs group2 {1}'.format(np.round(group1.likes.mean()),
                                          np.round(group2.likes.mean())))
print('-'*50, '\n')

print('dislikes Comparision\n')
print('group1 : {0} vs group2 {1}'.format(np.round(group1.dislikes.mean()),
                                          np.round(group2.dislikes.mean())))

likes Comparision

group1 : 62498.0 vs group2 74806.0
-------------------------------------------------- 

dislikes Comparision

group1 : 926.0 vs group2 866.0


0.030451351149508275

In [45]:
# 조회수대비 좋아요, 싫어요 비율의 평균비교

print('likes Rate Comparision\n')
print('group1 : {0} vs group2 {1}'.format(
    np.round((group1.likes.values / group1.view_count.values).mean(),5),
    np.round((group2.likes.values / group2.view_count.values).mean(),5)))
print('-'*50, '\n')

print('dislikes Rate Comparision\n')
print('group1 : {0} vs group2 {1}'.format(
    np.round((group1.dislikes.values / group1.view_count.values).mean(),5), 
    np.round((group2.dislikes.values / group2.view_count.values).mean(),5)))

likes Rate Comparision

group1 : 0.03289 vs group2 0.03045
-------------------------------------------------- 

dislikes Rate Comparision

group1 : 0.00059 vs group2 0.00042


### 아주 근소하게 Group1(인기동영상이 여러개인 채널)이
### Group2(인기동영상이 1개인 채널)보다 조회수, 좋아요수, 싫어요수, 좋아요비율, 싫어요비율이 약간 더 높다.